internal/host/plugin/repository_host_catalog_test.go

```diff
@@ -368,12 +369,736 @@ func TestRepository_CreateCatalog(t *testing.T) {
+				"three": "four",
+			})},
+			version:   2,
+			fieldMask: []string{"secrets.three"},
+			wantCheckFuncs: []checkFunc{
+				checkVersion(2), // no-op
+				checkUpdateCatalogRequestPersistedSecrets(map[string]interface{}{
+					"one": "two",
+				}),
+				checkUpdateCatalogRequestSecrets(map[string]interface{}{
+					"three": "four",
+				}),
+				checkSecrets(map[string]interface{}{
+					"one": "two",
+				}),
+				checkNumUpdated(db.NoRowsAffected),
+			},
+		},
+		{
+			name:        "delete secrets",
+			changeFuncs: []changeHostCatalogFunc{changeSecrets(map[string]interface{}{})},
+			version:     2,
+			fieldMask:   []string{"secrets"},
+			wantCheckFuncs: []checkFunc{
+				checkVersion(3),
+				checkUpdateCatalogRequestPersistedSecrets(map[string]interface{}{
+					"one": "two",
+				}),
+				checkUpdateCatalogRequestSecrets(map[string]interface{}{}),
+				checkSecretsDeleted(),
+				checkNumUpdated(1),
+				checkVerifyCatalogOplog(oplog.OpType_OP_TYPE_UPDATE),
+			},
+		},
+		{
+			name: "update name and secrets",
+			changeFuncs: []changeHostCatalogFunc{
+				changeName("foo"),
+				changeSecrets(map[string]interface{}{
+					"three": "four",
+				}),
+			},
+			version:   2,
+			fieldMask: []string{"name", "secrets.three"},
+			wantCheckFuncs: []checkFunc{
+				checkVersion(3),
+				checkUpdateCatalogRequestCurrentNameNil(),
+				checkUpdateCatalogRequestNewName("foo"),
+				checkUpdateCatalogRequestSecrets(map[string]interface{}{
+					"three": "four",
+				}),
+				checkName("foo"),
+				checkSecrets(map[string]interface{}{
+					"three": "four",
+				}),
+				checkNumUpdated(1),
+				checkVerifyCatalogOplog(oplog.OpType_OP_TYPE_UPDATE),
+			},
+		},
+	}
+
+	// Finally define a function for bringing the test subject catalog.
+	// This function also returns a function to clean up the catalog
+	// afterwards.
+	setupHostCatalog := func(t *testing.T, ctx context.Context) (*HostCatalog, func()) {
+		t.Helper()
+		require := require.New(t)
+
+		cat := TestCatalog(t, dbConn, projectScope.PublicId, testPlugin.GetPublicId())
+		// Set some (default) attributes on our test catalog
+		cat.Attributes = mustMarshal(map[string]interface{}{
+			"foo": "bar",
+		})
+
+		numCatUpdated, err := dbRW.Update(ctx, cat, []string{"attributes"}, []string{})
+		require.NoError(err)
+		require.Equal(1, numCatUpdated)
+
+		// Set up some secrets
+		cSecretProto := mustStruct(map[string]interface{}{
+			"one": "two",
+		})
+		cSecret, err := newHostCatalogSecret(ctx, cat.GetPublicId(), cSecretProto)
+		require.NoError(err)
+		scopeWrapper, err := dbKmsCache.GetWrapper(ctx, cat.GetScopeId(), kms.KeyPurposeDatabase)
+		require.NoError(err)
+		require.NoError(cSecret.encrypt(ctx, scopeWrapper))
+		cSecretQ, cSecretV := cSecret.upsertQuery()
+		secretsUpdated, err := dbRW.Exec(ctx, cSecretQ, cSecretV)
+		require.NoError(err)
+		require.Equal(1, secretsUpdated)
+
+		cleanupFunc := func() {
```

Comment:
t.Cleanup() takes a function and calls it once the *testing.T receiver's tests/subtests finish.  That way you don't have to pass the cleanup onto the caller.  Would that be helpful here?

Reply:
Will noodle on it for a little bit (not too long). It might add some idiomatic value but I think the workflow is clear enough that it's not 100% needed, and there's some other defers going on that I think would be good.

I think it depends on if cleanup is called after the parent finishes or if the subtest finishes, since the intention is to have it cleaned up after the subtest does.

Reply:
Think this will be fine. According to t.Cleanup():

https://pkg.go.dev/testing#T.Cleanup

> Cleanup registers a function to be called when the test (or subtest) and all its subtests complete. Cleanup functions will be called in last added, first called order. 

I've updated the code to use this.